### PR TITLE
Rewrite events.py so event & playbook models are well-defined 

### PR DESCRIPTION
**Pull request by noxasaxon**
Does not attempt to change any of the logic for turning an event into a playbook execution. Focuses on removing the EventBatch, EventCreator classes into Event classes with methods that use the event.

Simliarly, `create_events()` spells out more of whats going on by directly calling the necessary methods from event formatting all the way to playbook start. 

Breaking these methods up away from init() functions or nested in other methods will help as we decide whether all these steps will ultimately be necessary in `2.0.0`

- The only major feature improvement besides cleanup is that create_events() can now support a single event with `event["details"]` as a dict in addition to the previously supported format of multiple events with details as a list of dicts.

--------
Previous create_events() call flow:
1. `create_events` calls `EventBatch().create_events`
2. `EventBatch()` init ensures the event keys are the right types
3. `EventBatch().create_events` says for each event in details list:
   1.  format the event to a new structure and then call `EventCreator(new_event).create()`
   2.  `EventBatch().execute_playbook()` - step 7
4.  `EventCreator` init validates event key types
5.  `EventCreator` `.create()` calls `.deduplicate()` which interacts with DEDUP_TABLE and mutates the event
6.  `.create()` then formats the event and puts it into the EVENTS_TABLE
7.  `EventBatch().execute_playbook()` 
    1.  formats the playbook metadata (arn, investigation_id, input dict)
    2.  puts an item in the results table
    3.  starts stepFunctions with a subset of results item

After this PR:
1. create_events:
   1. generates a datetime for all events
   2. converts all events into an InitialEvent dataclass which checks types
   3. for each event in the list:
      1. feed the initialevent class into a completeevent class which generates metadata and methods surrounding the initialevent struct
      2. deduplicate
      3. put_item of event in events_table
      4. start_playbook
      5. save a message about whether that playbook started
   4. raise an error after all messages are started, if any of them failed to start

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [X] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: socless/events.py [hunk 1]
```diff
@@ -94,215 +142,211 @@ def dedup_hash(self):
...
+            InitialEvent(
+                details=event_details["details"],
+                created_at=event_details["created_at"],
+                event_type=event_details["event_type"],
+                playbook=event_details["playbook"],
+                data_types=event_details["data_types"],
+                event_meta=event_details["event_meta"],
+                dedup_keys=event_details["dedup_keys"],
+            )
+        )
```
Comment (ubaniabalogun) [inline socless/events.py:315]: See the other comment tagged #Ref-InitialEvent.
**tldr** Some of these values should be set to defaults reasonable defaults, as not all use-cases that call create_events pass all of them currently.

File: socless/events.py [hunk 2]
```diff
@@ -94,215 +142,211 @@ def dedup_hash(self):
...
+            playbook=str(self.event.playbook),
+            statemachinearn=playbook_arn,
+            execution_id=self.metadata.execution_id,
+            error="",
         )
-        stepfunctions = boto3.client("stepfunctions")
         try:
-            _ = stepfunctions.start_execution(
-                name=execution_id,
+            _ = stepfunctions_client.start_execution(
```
Comment (ubaniabalogun) [inline socless/events.py:270]: The unused assignment here is unneeded

File: socless/events.py [hunk 3]
```diff
@@ -94,215 +142,211 @@ def dedup_hash(self):
...
-            )
-            socless_log.info(
-                "Playbook execution started",
-                dict(
-                    meta,
-                    **{"statemachinearn": playbook_arn, "execution_id": execution_id},
-                ),
+                input=json.dumps(playbook_input_as_dict),
             )
+            socless_log.info("Playbook execution started", report.__dict__)
```
Comment (ubaniabalogun) [inline socless/events.py:275]: #question: is calling the `__dict__` method here the only way to get out the data we need?

File: socless/events.py [hunk 4]
```diff
@@ -94,215 +142,211 @@ def dedup_hash(self):
...
-    """Use the EventBatch class to create events
-    Args:
-        event_details (dict): The details of the events
-        context (obj): The Lambda context object
-    Returns:
-        dict containing the execution ids of the created events
-    """
-    event_batch = EventBatch(event_details, context)
-    return event_batch.create_events()
+def validate_event(event):
```
Comment (ubaniabalogun) [inline socless/events.py:335]: This function seems to be unused

File: socless/models.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+
+@dataclass
+class StateConfig:
+    Name: str
+    Parameters: dict
+
+
+@dataclass
+class SoclessContext:
```
Comment (ubaniabalogun) [inline socless/models.py:58]: #changerequest: Let's find a more descriptive name for this. 
Some recommendations
- PlaybookExecutionContext
- ExecutionContext
- SOClessPlaybookExecutionContext

The key take away here is this should reflect the scope of this Context Object

Comment (ubaniabalogun) [reply socless/models.py:58]: Also seems like there's some overlap between this object and the `PlaybookInput` object as well.

Is one a superset of the other? If so, the code should represent that as well

**Review by ubaniabalogun — CHANGES_REQUESTED** [review]

File: socless/events.py [hunk 6]
```diff
@@ -26,56 +29,101 @@
...
+    _cached_dedup_hash = dedup_hash
+    key = {"dedup_hash": _cached_dedup_hash}
+    dedup_mapping = dedup_table.get_item(Key=key).get("Item")
 
-        self.details = event_info.get("details", {})
+    if dedup_mapping:
+        try:
+            return dedup_mapping["current_investigation_id"]
+        except KeyError:
+            socless_log.warn(
```
[the remark this answers is not in the document]
Comment (noxasaxon) [reply socless/events.py:70]: this is the same code that was there previously, i didnt try to change any existing logic just refactor

File: socless/events.py (hunk 3 above)
Comment (noxasaxon) [reply socless/events.py:275]: converting a dataclass to a dict can be done two different ways (besides manually building the dict:
- __dict__ is fast but isn't recursive. if one of your dataclass attributes is another class/dataclass, its not gonna work correctly
- dataclasses.asdict(object) is ~ 10x slower but works recursively.

File: socless/events.py [hunk 7]
```diff
@@ -26,56 +29,101 @@
...
-                    "Error: Supplied 'created_at' field is not ISO8601 millisecond-precision string, shifted to UTC"
-                )
+def get_investigation_id_from_dedup_table_or_raise_not_found(dedup_hash: str) -> str:
+    """Check dedup_table for an investigation_id using the dedup_hash.
+    Raises:
+        SoclessNotFoundError if dedup_table item is malformed or doesnt exist.
+    Logs a warning if a dedup_hash is found but doesn't contain any investigation_id.
+    """
+    _cached_dedup_hash = dedup_hash
+    key = {"dedup_hash": _cached_dedup_hash}
```
[the remark this answers is not in the document]
Comment (noxasaxon) [reply socless/events.py:63]: vestigial code i brought over from eventcreator, i'll remove it

File: socless/events.py [hunk 8]
```diff
@@ -26,56 +29,101 @@
...
-        self.event_info = event_info
+def get_playbook_arn(playbook_name, lambda_context):
+    return "arn:aws:states:{region}:{accountid}:stateMachine:{stateMachineName}".format(
+        region=os.environ["AWS_REGION"],
+        accountid=lambda_context.invoked_function_arn.split(":")[4],
+        stateMachineName=playbook_name,
+    )
+
+
+def setup_results_table_for_playbook(
```
[the remark this answers is not in the document]
Comment (noxasaxon) [reply socless/events.py:40]: changed to setup_results_table_for_playbook_execution

File: socless/events.py [hunk 9]
```diff
@@ -26,56 +29,101 @@
...
+
+@dataclass
+class InitialEvent:
+    created_at: str
+    event_type: str
+    playbook: Optional[str]
+    details: dict
+    data_types: dict
+    event_meta: dict
+    dedup_keys: list
```
[the remark this answers is not in the document]
Comment (noxasaxon) [reply socless/events.py:108]: added defaults to data_types, event_meta, dedup_keys

File: socless/events.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ubaniabalogun) [reply socless/events.py:69]: I guess I past Ubani needs to answer that question then 🤣

**Review by ubaniabalogun — APPROVED** [review]
